### PR TITLE
*: fix the lower bound when converting numbers less than 0 to unsigned integers

### PR DESCRIPTION
**Pull request by exialin**
### What problem does this PR solve? 
Fix issue #6360 so that when `set sql_mode = ''`, values less than 0 will be clipped to 0 for unsigned integer types. Besides, some other cases are also fixed.

- Insert

MySQL:
```sql
mysql> create table t(a bigint unsigned);
Query OK, 0 rows affected (5.01 sec)

mysql> set sql_mode='';
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql> insert into t value(-1.111);
Query OK, 1 row affected, 1 warning (0.16 sec)

mysql> insert into t value('-1.111');
Query OK, 1 row affected, 1 warning (0.17 sec)

mysql> select * from t;
+------+
| a    |
+------+
|    0 |
|    0 |
+------+
2 rows in set (0.00 sec)
```

Previously in TiDB:
```sql
tidb> create table t(a bigint unsigned);
Query OK, 0 rows affected (0.01 sec)

tidb> set sql_mode='';
Query OK, 0 rows affected (0.01 sec)

tidb> insert into t value(-1.111);
Query OK, 1 row affected, 1 warning (0.00 sec)

tidb> insert into t value('-1.111');
Query OK, 1 row affected, 1 warning (0.00 sec)

tidb> select * from t;
+----------------------+
| a                    |
+----------------------+
| 18446744073709551615 |
|                 NULL |
+----------------------+
2 rows in set (0.00 sec)
```

- Update

See issue #8623. If the problem described in that issue is fixed, the wrong behaviour can be fixed by rountines of this pull request (test cases should be added later).

- Load data

Data file:
```shell
$ cat a.csv 
-1
-18446744073709551615
-18446744073709551616
```

MySQL:
```sql
mysql> create table t(a bigint unsigned);
Query OK, 0 rows affected (0.32 sec)

mysql> load data local infile '/home/exialin/Test/6360/a.csv' into table t;
Query OK, 3 rows affected, 3 warnings (0.07 sec)
Records: 3  Deleted: 0  Skipped: 0  Warnings: 3

mysql> select * from t;
+----------------------+
| a                    |
+----------------------+
|                    0 |
|                    0 |
| 18446744073709551615 |
+----------------------+
3 rows in set (0.00 sec)
```

Previously in TiDB:
```sql
tidb> load data local infile '/home/exialin/Test/6360/a.csv' into table t;
Query OK, 1 row affected, 1 warning (0.01 sec)

tidb> select * from t;
+------+
| a    |
+------+
| NULL |
| NULL |
| NULL |
+------+
3 rows in set (0.00 sec)
```

After this fix:
```sql
mysql> select * from t;
+------+
| a    |
+------+
|    0 |
|    0 |
|    0 |
+------+
```

**Question**: is the third row of MySQL reasonable?

- Alter table

TiDB doesn't support the following `alter table` query currently. Just post for completeness.

MySQL:
```sql
mysql> create table t (a bigint);
Query OK, 0 rows affected (0.41 sec)

mysql> insert into t value (-1);
Query OK, 1 row affected (0.11 sec)

mysql> alter table t modify column a bigint unsigned;
ERROR 1264 (22003): Out of range value for column 'a' at row 1
mysql> set sql_mode='';
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql> alter table t modify column a bigint unsigned;
Query OK, 1 row affected, 1 warning (1.15 sec)
Records: 1  Duplicates: 0  Warnings: 1

mysql> select * from t;
+------+
| a    |
+------+
|    0 |
+------+
1 row in set (0.00 sec)
```

### What is changed and how it works?

MySQL [doc](https://dev.mysql.com/doc/refman/5.7/en/out-of-range-and-overflow.html) says:
> If no restrictive modes are enabled, MySQL clips the value to the appropriate endpoint of the column data type range and stores the resulting value instead.

> When strict SQL mode is not enabled, column-assignment conversions that occur due to clipping are reported as warnings for ALTER TABLE, LOAD DATA INFILE, UPDATE, and multiple-row INSERT statements.

So this makes the result different from other conversions like `CAST()` function. When the number to be inserted is less than 0, it should be clipped to 0.

**TODO**: the warning message has not been fixed yet. It would be better to fix them in other pull requests.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported variable/fields change

Side effects

 - Increased code complexity. It seems that the implementation is a bit unclean.

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8544)

### CODE REVIEW

**Comment by exialin** [comment]
Currently `make dev` will fail:

```
OOPS: 113 passed, 1 FAILED
--- FAIL: TestT (0.44s)
FAIL
coverage: 77.5% of statements
FAIL	github.com/pingcap/tidb/types	0.912s
```

But it doesn't show which line caused the error. I tried to pick them out but still missed some. So is there any way to locate the line?

**Comment by tiancaiamao** [comment]
```
6,14c6,14
<     └─IndexJoin_24	0.00	root	inner join, inner:IndexLookUp_23, outer key:gad.aid, inner key:test.dd.aid, other cond:eq(test.dd.ip, gad.ip), gt(test.dd.t, gad.t)
<       ├─IndexLookUp_23	0.00	root	
<       │ ├─IndexScan_20	10.00	cop	table:dd, index:aid, dic, range: decided by [gad.aid gad.ip], keep order:false, stats:pseudo
<       │ └─Selection_22	0.00	cop	eq(test.dd.bm, 0), eq(test.dd.pt, "android"), gt(test.dd.t, 1478143908)
<       │   └─TableScan_21	10.00	cop	table:dd, keep order:false, stats:pseudo
<       └─IndexLookUp_33	0.00	root	
<         ├─IndexScan_30	3333.33	cop	table:gad, index:t, range:(1478143908,+inf], keep order:false, stats:pseudo
<         └─Selection_32	0.00	cop	eq(gad.bm, 0), eq(gad.pt, "android")
<           └─TableScan_31	3333.33	cop	table:st, keep order:false, stats:pseudo
\ No newline at end of file
---
>     └─IndexJoin_24	0.00	root	inner join, inner:IndexLookUp_23, outer key:gad.aid, inner key:test.dd.aid, other cond:eq(gad.ip, test.dd.ip), gt(test.dd.t, gad.t)
>       ├─IndexLookUp_33	0.00	root	
>       │ ├─IndexScan_30	3333.33	cop	table:gad, index:t, range:(1478143908,+inf], keep order:false, stats:pseudo
>       │ └─Selection_32	0.00	cop	eq(gad.bm, 0), eq(gad.pt, "android")
>       │   └─TableScan_31	3333.33	cop	table:st, keep order:false, stats:pseudo
>       └─IndexLookUp_23	0.00	root	
>         ├─IndexScan_20	10.00	cop	table:dd, index:aid, dic, range: decided by [gad.aid gad.ip], keep order:false, stats:pseudo
>         └─Selection_22	0.00	cop	eq(test.dd.bm, 0), eq(test.dd.pt, "android"), gt(test.dd.t, 1478143908)
>           └─TableScan_21	10.00	cop	table:dd, keep order:false, stats:pseudo
\ No newline at end of file
```

The order of explain result  of the `IndexJoin` is not stable? @winoros

**Comment by winoros** [comment]
We're checking it.

File: sessionctx/stmtctx/stmtctx.go [hunk 1]
```diff
@@ -61,6 +62,10 @@ type StatementContext struct {
 	PadCharToFullLength    bool
 	BatchCheck             bool
 	InNullRejectCheck      bool
+	// As for insert, update, alter table and load data infile statements, when not in strict SQL mode,
+	// the value should be clipped to 0 for unsigned integer types.
+	// see https://dev.mysql.com/doc/refman/5.7/en/out-of-range-and-overflow.html
+	ClipToZero bool
```
[the remark this answers is not in the document]
Comment (exialin) [reply sessionctx/stmtctx/stmtctx.go]: Thanks for your suggestion. Now I use a function instead to keep the `if` condition clear.

**Comment by exialin** [comment]
After more tests, I found there are two special cases hard to resolve:

1. the value is slightly larger than max `uint64`

For example:
```sql
tidb> create table t(a bigint unsigned);
Query OK, 0 rows affected (0.03 sec)

tidb> insert into t value (18446744073709551616);
Query OK, 1 row affected (0.00 sec)


tidb> select * from t;
+---------------------+
| a                   |
+---------------------+
| 9223372036854775808 |
+---------------------+
1 row in set (0.00 sec)

tidb> insert into t value (18446744073709554616);
ERROR 1264 (22003): Out of range value for column 'a' at row 1
```
Here `18446744073709551616 (2^64)` should cause an error, but it is truncated to an incorrect value. Because `18446744073709551616` is treated as a float, so it goes to `types.ConvertFloatToUint`. Due to precision loss of float, it doesn't cause an overflow error.

2. when not in strict mode, values in certain range still cause errors instead of warnings

For example:
```sql
tidb> set sql_mode='';
Query OK, 0 rows affected (0.01 sec)

tidb> insert into t value (-9223372036854775809);
ERROR 1264 (22003): Out of range value for column 'a' at row 1
tidb> insert into t value (-18446744073709551615);
ERROR 1264 (22003): Out of range value for column 'a' at row 1
```
To be compatible with MySQL, here values in **`[-18446744073709551615, -9223372036854775809]`** should be clipped to 0. This is a rather special case that only happens for values in this range. Take a look at `insert_common.go/evalRow`:
https://github.com/pingcap/tidb/blob/808eab19115d81a2940bca87e287dd7e58218c75/executor/insert_common.go#L247-L254

Other integer literals are evaluated correctly in `expr.Eval`, but values in this range will cause errors, hence not going to `table.CastValue`. For a value like `-9223372036854775809`, here `expr` is a `ScalarFunction` which represents `builtinUnaryMinusIntSig`, which means it is treated as a `Minus` expression and an integer. Since `9223372036854775809` overlows `int64`, an error will be returned.

I tried to do some workaround like making `ScalarFunction.Eval` return 0 and appending the error in warning queue, but it would fail other test cases and it isn't a clean way. Perhaps the best way is to evaluate values in this range as float.

@zz-jason Do you have any suggestion for these two cases?

File: sessionctx/stmtctx/stmtctx.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (exialin) [reply sessionctx/stmtctx/stmtctx.go]: @zz-jason PTAL

**Comment by zz-jason** [comment]
@exialin Can not find out any easy way to fix them either. Could you file a github issue to record these two cases so that one day we may be able to fix it?

**Comment by exialin** [comment]
@zz-jason OK. Maybe after changes in this PR are merged? The two cases are somewhat related with this PR.

**Comment by zz-jason** [comment]
> @zz-jason OK. Maybe after changes in this PR are merged? The two cases are somewhat related with this PR.

OK

**Comment by zz-jason** [comment]
LGTM

File: executor/write_test.go [hunk 2]
```diff
@@ -1920,6 +1941,28 @@ func (s *testSuite2) TestLoadDataIgnoreLines(c *C) {
...
+	tk.MustExec("CREATE TABLE load_data_test (a bigint unsigned);")
+	tk.MustExec("load data local infile '/tmp/nonexistence.csv' into table load_data_test")
+	ctx := tk.Se.(sessionctx.Context)
+	ld, ok := ctx.Value(executor.LoadDataVarKey).(*executor.LoadDataInfo)
+	c.Assert(ok, IsTrue)
+	defer ctx.SetValue(executor.LoadDataVarKey, nil)
+	c.Assert(ld, NotNil)
+	tests := []testCase{
+		{nil, []byte("-1\n-18446744073709551615\n-18446744073709551616\n"), []string{"0", "0", "0"}, nil, "Records: 3  Deleted: 0  Skipped: 0  Warnings: 3"},
+		// manual test can pass but here will fail
```
Comment (XuHuaiyu) [inline executor/write_test.go]: I do not understand what does this comment mean.
Do you mean that `load data local infile` in MySQL client runs well?

File: types/convert.go [hunk 3]
```diff
@@ -124,9 +128,12 @@ func ConvertUintToUint(val uint64, upperBound uint64, tp byte) (uint64, error) {
 }
 
 // ConvertFloatToUint converts a float value to an uint value.
-func ConvertFloatToUint(fval float64, upperBound uint64, tp byte) (uint64, error) {
+func ConvertFloatToUint(sc *stmtctx.StatementContext, fval float64, upperBound uint64, tp byte) (uint64, error) {
 	val := RoundFloat(fval)
 	if val < 0 {
+		if sc != nil && sc.ShouldClipToZero() && val < 0 {
```
Comment (XuHuaiyu) [inline types/convert.go]: We do not need to check `val<0` again.

File: expression/builtin_cast.go [hunk 4]
```diff
@@ -464,7 +464,7 @@ func (b *builtinCastIntAsRealSig) evalReal(row chunk.Row) (res float64, isNull b
 		res = 0
 	} else {
 		var uVal uint64
-		uVal, err = types.ConvertIntToUint(val, types.UnsignedUpperBound[mysql.TypeLonglong], mysql.TypeLonglong)
+		uVal, err = types.ConvertIntToUint(nil, val, types.UnsignedUpperBound[mysql.TypeLonglong], mysql.TypeLonglong)
```
Comment (XuHuaiyu) [inline expression/builtin_cast.go]: Why not pass `sc` here?

File: executor/write_test.go (hunk 2 above)
Comment (exialin) [reply executor/write_test.go]: I've fixed it. Sorry for the confusion.

File: expression/builtin_cast.go (hunk 4 above)
Comment (exialin) [reply expression/builtin_cast.go]: @XuHuaiyu To my understanding, this function is used for functions like `CAST`, and there is no difference whether we set `sql_mode=''`:
```sql
mysql> select cast(-1 as unsigned);
+----------------------+
| cast(-1 as unsigned) |
+----------------------+
| 18446744073709551615 |
+----------------------+
1 row in set (0.55 sec)

mysql> set sql_mode='';
Query OK, 0 rows affected, 1 warning (0.00 sec)

mysql> select cast(-1 as unsigned);
+----------------------+
| cast(-1 as unsigned) |
+----------------------+
| 18446744073709551615 |
+----------------------+
1 row in set (0.00 sec)
```
So no need to pass `sc` to check.

**Comment by XuHuaiyu** [comment]
hi, @exialin Will the other 2 comments be addressed?

File: expression/builtin_cast.go (hunk 4 above)
Comment (exialin) [reply expression/builtin_cast.go]: Reasonable. I was using this `nil` to differ `CAST` from this case. It turns out that there is no need to do that. Nice catch :-)

**Comment by XuHuaiyu** [comment]
The following sql failed, it seems to be unexpected.
``` sql
CREATE TABLE `table1_int_autoinc` (
`col_int_unsigned_not_null_unique` int(10) unsigned NOT NULL,
UNIQUE KEY `col_int_unsigned_not_null_unique` (`col_int_unsigned_not_null_unique`)
) ;
DELETE FROM `table1_int_autoinc` WHERE `col_int_unsigned_not_null_unique` = -1438187520 ;
```

**Comment by exialin** [comment]
@XuHuaiyu CI test fails when executing ``UPDATE `table10_int_autoinc` SET `col_int` = 335740928 WHERE `col_int_unsigned_not_null_unique` = -798228480;``, because `-798228480` is incorrectly converted.
I'm wondering which is better: fix it in this PR or fix it together with #8623 later. Or fix #8623 in this PR...

File: sessionctx/stmtctx/stmtctx.go [hunk 5]
```diff
@@ -379,3 +381,24 @@ func (sc *StatementContext) GetExecDetails() execdetails.ExecDetails {
...
+// see https://dev.mysql.com/doc/refman/5.7/en/out-of-range-and-overflow.html
+func (sc *StatementContext) ShouldClipToZero() bool {
+	// TODO: Currently altering column of integer to unsigned integer is not supported.
+	// If it is supported one day, that case should be added here.
+	return sc.InInsertStmt || sc.InLoadDataStmt
+}
+
+// ShouldIgnoreError indicates whether we should ignore the error when type conversion overflows,
+// so we can leave it for further processing like clipping values less than 0 to 0 for unsigned integer types.
+func (sc *StatementContext) ShouldIgnoreError() bool {
```
Comment (jackysp) [inline sessionctx/stmtctx/stmtctx.go]: It is better to clarify what kind of error in the function name.

Comment (exialin) [reply sessionctx/stmtctx/stmtctx.go]: Modified as you suggested.

**Comment by XuHuaiyu** [comment]
It will be better to fix #8623 in an individual PR.

**Comment by zz-jason** [comment]
@exialin Could you cherry-pick this commit to release-2.1 and release-2.0 branch?

**Comment by exialin** [comment]
@zz-jason OK, I'll do it later and file a issue about the unresolved cases. Thanks for the review from all of you!

**Pull request merged**